### PR TITLE
Support method deletion in `static_methods()` invalidation.

### PR DESCRIPTION
**Pull request by NHDaly**
Add all matching method instances as backedges to `static_methods()`, so
that deleting a method (which apparently only invalidates functions that
inlined that method), will still invalidate our `static_methods()`
function.

Adds test for method deletion. :)

<s>
Note that there is a problem with detecting covering methods when the
provided type includes Varargs, so we currently oversubscribe to more
MethodInstances than we need to. :/
</s>

### CODE REVIEW

File: src/Tricks.jl [hunk 1]
```diff
@@ -70,10 +70,20 @@ static_methods(@nospecialize(f)) = static_methods(f, Tuple{Vararg{Any}})
...
-    ci.edges = Core.Compiler.vect(mt, Tuple{Vararg{Any}})
+    mt_edges = Core.Compiler.vect(mt, Tuple{Vararg{Any}})
+
+    world = typemax(UInt)
+    method_insts = Core.Compiler.method_instances(f.instance, T, world)
+    # TODO: UGH except this actually *doesn't work* with Tuple{Vararg{Any}}.... So currently
+    #       we are over-subscribing to invalidations, but that's better than incorrectness.
+    # ftype = Tuple{f, T.parameters...}
+    # covering_method_insts = [mi for mi in method_insts if ftype <: mi.def.sig]
+    covering_method_insts = method_insts
```
[the remark this answers is not in the document]
Comment (NHDaly) [reply src/Tricks.jl:81]: Ah, sorry, i should have added more detail to my message. The problem is that (somehow) `methods()` has support for signatures that contain `Vararg{}`, and returns methods that (might) match the varags:
```julia
julia> foo(x, y...) = 2
foo (generic function with 1 method)

julia> methods(foo, Tuple{Vararg{Any}})
# 1 method for generic function "foo":
[1] foo(x, y...) in Main at REPL[1]:1
```

Even though **there might not actually be a match there!**

-----

But okay, now that i'm rereading the above, i'm realizing that the answer is that indeed: **We shouldn't be filtering down the list at all.**

If any method in the list of methods returned is changed, we want to invalidate this result, even if one of those methods wasn't an "actual match" for our provided type tuple.

So i think the "hacky" behavior that i implemented above is actually 100% correct! :) I'm going to clean up the comments with my newfound understanding and then this should be ready to go 😎

[the remark this answers is not in the document]
Comment (NHDaly) [reply src/Tricks.jl]: I've added some tests that exercise the case i was talking about, and hopefully that helps to make this clearer:
https://github.com/oxinabox/Tricks.jl/pull/34/files#diff-3b9314a6f9f2d7eec1d0ef69fa76cfabafdbe6d0df923768f9ec32f27a249c63R121

**Comment by NHDaly** [comment]
This should be ready to go now. :) Give it a look when you get some time, and i'll Squash+merge it after that :)

File: src/Tricks.jl [hunk 2]
```diff
@@ -70,11 +70,26 @@ static_methods(@nospecialize(f)) = static_methods(f, Tuple{Vararg{Any}})
...
+    # are defined, it recompiles the function.
+    mt_edges = Core.Compiler.vect(mt, Tuple{Vararg{Any}})
+
+    # We want to add an edge to _every existing method instance_, so that
+    # the deletion of any one of them will trigger recompilation of the function.
+    world = typemax(UInt)
+    method_insts = Core.Compiler.method_instances(f.instance, T, world)
+    covering_method_insts = method_insts
+
+    return [mt_edges..., covering_method_insts...]
```
Comment (oxinabox) [inline src/Tricks.jl:90]: Maybe prefer:
```suggestion
    return [mt_edges; covering_method_insts]
```
or 
```suggestion
    return vcat(mt_edges, covering_method_insts)
```

**Review by oxinabox — APPROVED** [review]
Bump the version, then merge and tag when you are happy

File: src/Tricks.jl (hunk 2 above)
Comment (rai-nhdaly) [reply src/Tricks.jl:90]: haha huh i really don't like the first option. Is there a difference semantically, or just syntactically?

I'll accept the second suggestion 👍

**Comment by NHDaly** [comment]
i'm gonna tag a release now too :) Thanks lyndon! :)

**Pull request merged**